### PR TITLE
Add the possibility to change the codec of Framed

### PR DESCRIPTION
**Pull request by mredlek**
I added a function to replace the codec of the Framed object.

My usecase for this are websockets. It starts with a simple line-based codec (or http-codec). After a connection upgrade to websockets are negotiogated, the frames change to a different codec.

If the codec change happends to occur after the first websocket was received by the tcp-stream, I am afraid those bytes are lost. So I added a simpel function to prevent this race condition by replacing the existing codec with a new codec and reuse the buffers.

### CODE REVIEW

**Comment by carllerche** [comment]
Thanks for the PR. I believe that the main blocker for doing this today is that there is no way to get the buffer out from `Framed`. I think that I would rather see a way to unpack a `Framed` (get the codec & buffer) then create a new one.

Thoughts @alexcrichton?

**Comment by mredlek** [comment]
I did thought about that. For my purposes it will also work perfectly. I did choose this method because I feeled it did hide implementation details better (such as the type of buffer that is used).

If the buffers can be split out just as the transport, that also a create method needs to be added to created a Framed-object with buffers.

**Comment by alexcrichton** [comment]
Yeah I think I'd lean the same way as @carllerche allowing the ability to extract the buffer (and likely the underlying I/O object as well) and then allowing transformation externally via recreation of a new encoder/decoder.

**Comment by carllerche** [comment]
@alexcrichton what do you think about adding `into_parts(self) -> (T, BytesMut)`?

**Comment by alexcrichton** [comment]
👍

**Comment by carllerche** [comment]
@mredlek Do you have an interest to update the PR to provide `into_parts` instead?

**Comment by mredlek** [comment]
Sure. I will probably finish that next weekend.

I think the signature should be `into_parts(self) -> (T, BytesMut, BytesMut)` (stream, read buffer, write buffer).

Any suggestion how to name the creation function? `framed_from_parts`?

**Comment by carllerche** [comment]
For `Framed`, one option would be to have `into_parts` return `(FramedWrite, read buf)`. At three fields, it may worth creating a `FramedParts` struct or something...

Thoughts @alexcrichton

**Comment by alexcrichton** [comment]
Hm  yeah I don't think we'll want to return a tuple with two `BytesMut` entries in it, we'll want something to disambuiguate that. In that sense I'd personally at least like to optimize for readability here, so I wouldn't mind having a "throwaway struct" like `std::process::Output` which just contains a bunch of fields, I like `FramedParts` as the name.

For a constructor I'd recommend `Framed::from_parts` or something like that, I don't think this function would need to go on the `AsyncRead` trait directly.

**Comment by mredlek** [comment]
Do the fields in `FieldsPart` need to be public?

By the way, do you have any interest in a struct + trait to convert a transport of type A to a new transport of type B? I think the (Augmenting the transport)[https://tokio.rs/docs/going-deeper-tokio/transports/#augmenting-transport] can be generalized and made simpler for users.

**Comment by alexcrichton** [comment]
Yeah I was thinking it'd just have public fields for destruturing.

In theory transports are just sinks/streams, right? (which are easy to interconvert between types)

**Comment by carllerche** [comment]
@mredlek as Alex said, you should be able to `map` the transport... of course `Sink`'s combinators are a bit less obvious...

**Comment by mredlek** [comment]
The case I am working on is a service that needs json rpc via websockets. It first get a websocket connection (in a `Framed` object), then convert that into a transport that handles the websocket protocol (sends back ping, combines partial frames to a single frame), then it converts into a transport that parses the content as Json. Then in the next transport JsonRpc is handled, binding send request to incoming response and so on. I currently use a base trait for all the middleware conversion transports. It is currently not generic enough (I am rewriting it), but I might get generic enough for reuse.

Basically it let the user convert a simple incoming item into a list of items to be send down the stream and items to be send to the inner sink.

**Comment by mredlek** [comment]
This request is now obsoleted by pull request #45.

**Comment by belltoy** [comment]
I think not only the buffers(aka Framedparts) but also the codec and the underlying I/O are needed to be unpack. 

Generally, the codec keeps its state. To recreate the Framed object, it needs its underlying I/O, buffers, and the codec as well in some cases, or a new codec in other cases. So we can replace any of them if needed.

In my use case, I need to dispatch underlying I/O to different threads, we need get get the buffers and codec back, to recreate the framed object in the other thread.

I think an `into_codec(self) -> U` method and an `into_all(self) -> (T, U, FramedParts)` method should be needed of the Framed struct.

@carllerche @alexcrichton

**Comment by alexcrichton** [comment]
Ah unfortunately this is published now and stable, but @belltoy want to open an issue to track that?
